### ISSUE TITLE
Issue evaluation function

### ISSUE THREAD

**Issue by ofekih** [josdejong/mathjs#486]
When evaluating: (-1)^(-2/3), it returns a complex answer, when in fact it should simply return 1. the -1 should be squared before being cube rooted, not vice versa.

**Comment by FSMaxB** [comment]
I'm not a mathematician, but wolfram alpha also calculates a complex result: http://www.wolframalpha.com/input/?i=%28-1%29^%28-2%2F3%29 although [Wikipedia](https://en.wikipedia.org/wiki/Exponentiation#Rational_exponents) and [wolfram alpha](http://www.wolframalpha.com/input/?i=%28-27%29^%282%2F3%29) seem to disagree with each other.

This is definitely worth looking into.

**Comment by FSMaxB** [comment]
When taking the nth root of a negative number, there are n different complex solutions. This might be related to this problem.

**Comment by ofekih** [comment]
Yes, but since the squaring should be done first, there shouldn't be a
negative number

**Comment by ofekih** [comment]
Also: (-5)^(1/3)
Evaluates to a non-real solution, when a perfectly real one exists

**Comment by FSMaxB** [comment]
In case of a negative, mathjs currently does this:

```js
function _powComplex (x, y) {
    // complex computation
    // x^y = exp(log(x)*y) = exp((abs(x)+i*arg(x))*y)
    // TODO: we can optimize this as we know x and y are Complex
    //   expComplex      = exp.signatures['Complex,Complex']
    //   multiplyComplex = multiply.signatures['Complex,Complex']
    //   logComplex      = log.signatures['Complex,Complex']
    //   return expComplex(multiplyComplex(logComplex(x), y));
    return exp(multiply(log(x), y));
  }
```

Any idea how to improve this?

@josdejong: What is your take on this.

It might actually make sense to return more than one result. But that would most certainly be a breaking change.

**Comment by ofekih** [comment]
Yes, returning multiple solutions would be a way to do it, but the easiest
way to it it would be to check for a real solution and then return that if
it exists. I'm no mathematician either, but I'll look into it a bit.

On Oct 23, 2015 1:19 AM, "Max Bruckner" <notifications@github.com> wrote:
>
> In case of a negative, mathjs currently does this:
>
> function _powComplex (x, y) {
>     // complex computation
>     // x^y = exp(log(x)*y) = exp((abs(x)+i*arg(x))*y)
>     // TODO: we can optimize this as we know x and y are Complex
>     //   expComplex      = exp.signatures['Complex,Complex']
>     //   multiplyComplex = multiply.signatures['Complex,Complex']
>     //   logComplex      = log.signatures['Complex,Complex']
>     //   return expComplex(multiplyComplex(logComplex(x), y));
>     return exp(multiply(log(x), y));
>   }
>
> Any idea how to improve this?
>
> @josdejong: What is your take on this.
>
> It might actually make sense to return more than one result. But that
would most certainly be a breaking change.

**Comment by ericman314** [comment]
Matlab has a function `nthroot` that returns the real root, like this: `nthroot(-27, 3)` which returns `-3` instead of `1.5000000000000004 + 2.598076211353316i` as mathjs does. We could add this function.

We could also add, for completeness, a function `nthroots` that returns _all_ roots, including the real and complex ones. I'm not sure if matlab has this function or not.

@FSMaxB is right; evaluating `x^y` when `x` is negative and `y` is any real number has multiple possible values. The principal value is usually chosen so as to make the imaginary part of `log(x)` during the evaluation of `exp(multiply(log(x), y))` between -pi and pi:

```js
  function _logComplex(x) {
    return new type.Complex (
        Math.log(Math.sqrt(x.re * x.re + x.im * x.im)),
        Math.atan2(x.im, x.re)      // Imaginary part is between -pi and pi
    );
  }
```

**Comment by josdejong** [comment]
There are indeed three (complex) solutions for calculations with cubic roots like `(-1)^(-2/3)`. Normally, these calculations return the *prinicipal root*, which sometimes but not always is equal to the *real root* which you may prefer to get back.

The function [`math.cbrt(x [, allRoots])`](http://mathjs.org/docs/reference/functions/cbrt.html) has an optional second argument to return all three complex results. We could extend `nthRoot` to handle the special case where `root` is 3, and allow an optional third parameter `allRoots` like in `cbrt`. We could also extend `pow` to handle cases where the exponent is a multiple of 3 (also add an optional argument `allRoots`). I'm a bit afraid though that we run into issues with round-off errors if we try to match multiples of `1/3`. So maybe we should restrict that to the case where the exponent is a `Fraction` (without round-off errors).

**Comment by FSMaxB** [comment]
This not only applies to `1/3` but also to `1/n` in general, where `n` is an integer. Then there are `n` solutions if I understand it correctly.

**Comment by ofekih** [comment]
Yes, this makes sense. Thanks a lot for taking interest all of you!

I know I'm asking for much, but a real number mode would be very useful for
me (one that returns either a real solution or false). By the way, I'm
using math.js to make a simple grapher, and I'm worried for functions like
x^1/3. the-ofek-foundation.github.io/Grapher

**Comment by franklinyu** [comment]
For most people it makes sense to return the principal value. If you do want an additional "real mode", it *should not* be default, but may either be activated in the scope, like
```javascript
var scope = { realmode:true };
math.eval('(-8)^(-2/3)', scope); // -> 1/4
```
or, with a new function to find real root:
```javascript
math.eval('realrt(-1,3)^(-2)'); // -> 1/4
```

Personally I prefer the second; it is easier to implement, and it makes sense mathematically. It makes sure that the exponent is a rational number by factoring out the denominator.

**Comment by ericman314** [comment]
A real mode option for a grapher would definitely be useful. Google's calculator seems to have some logic built in to determine whether the user wants the real value or principal value. From what I can see, if the exponent is approximately the inverse of an integer, Google gives the real value; otherwise, it gives the principal value.

From Google Calculator:
```
(-5)^(1/3)
-1.70997594668

(-5)^(0.3333)
0.855097171 + 1.48071363 i

(-5)^(0.333333333333333)
0.854987973 + 1.48088261 i

(-5)^(0.3333333333333333)
-1.70997594668

(-5)^(2/3)
-1.46200887 + 2.53227364 i
```

In my opinion, the default behavior should always be to return the principal value. If the `realmode` option is set, then the `^` operator should return a real value if the exponent is approximately the inverse of an odd integer, and either the principal value or `NaN` otherwise.

Unfortunately, to make it work for an exponents such as `2/3`, which is first evaluated to `0.6666666...` we would have to find a sufficiently close [rational approximation](https://en.wikipedia.org/wiki/Continued_fraction#Best_rational_approximations) to recover the `2/3`. If the denominator is odd, then a real root exists! On a related note, this might also be a neat way to construct a `Fraction` type from a real value.

**Comment by franklinyu** [comment]
Well, the reason why I prefer a new function is exactly to avoid the script to determine the approximation, since it seems heavy burden for a script to do so while being trivial for user to split numerator and denominator. I have no idea about the complexity of calculating things like `(-17)^(29/137)`; it would certainly be a good idea if the calculation is affordable for a common explorer.

**Comment by ericman314** [comment]
I didn't realize before that [fraction.js](https://github.com/infusion/Fraction.js) already has the ability to create a fraction from a double value using something called Farey sequences, using a sort of binary search to find the closest fraction with denominator less than some chosen number. So we wouldn't need to code that ourselves. The overhead of converting to a fraction would only occur when the user explicitly asks for it by setting the `realmode` option. If users are really concerned with the overhead they can still use the `nthRoot` function: `nthRoot(-17, 137) ^ 29 = -1.8216292479175`

**Comment by ericman314** [comment]
I've put together an implementation of the `realmode` option. See #525. Can a few of you check it out and make sure it is mathematically correct?

**Comment by josdejong** [comment]
Thanks for implementing this experiment Eric! Looks good. It's cool to get a real value back (if existing). 

I'm still a bit in doubt about what form we should offer this functionality. If I'm not mistaken we're talking about 2 functions: `pow` and `nthRoot`. I think there are three options:

1. A global option like `realmode`, used internally by `pow` and `nthRoot`.
2. Separate functions: `realpow(x, y)` and `realNthRoot(x, root)`
3. An optional extra argument: `pow(x, y [, real=false])` and `nthRoot(x, root [, real=false])`.

Since this is an option very specific for just two functions, I'm not in favor for creating a global option for this, and rather prefer a separate function or optional argument. I think @franklinyu likes option (2) most. I like both (2) and (3). @ericman314, @FSMaxB, and @ofekih what do you think?

**Comment by ericman314** [comment]
There's also the `^` operator. A global option does seem like overkill but what other option is there for `^`? Could the real result behavior be appropriate to use with `^` if and only if `predictable` is set?

**Comment by josdejong** [comment]
Ah, that's a good point. There are two options then: implement a global option, or accept that the `^` operator cannot be used for real outputs, and if you want that you have to use the separate functions or pass this extra argument (options 2 and 3).

About the global option: If we go for that, I think it may be better to create a separate option like `realmode` instead of reusing the existing option `predictable`. The option `predictable` sounds similar to what we need for realmode, but you make it impossible to work in non-predictable output mode in combination with getting real output for `pow`. I think these should be two independently configurable options. @ericman314 do you have a strong preference for using a global option here?

**Comment by ericman314** [comment]
I think it's important to support the behavior for `^`, especially considering the use case of @ofekih's grapher. I don't like the idea of adding another global option just for a few functions though. If we do add the new option, it might make more intuitive sense to the user if it disables complex numbers entirely, so `sqrt(-1)` returns `NaN`. In this way it is similar to `predictable` like you said, but not exactly, since it only suppresses the output of complex numbers. I think this would be acceptable. 

Typically, when you pick up a calculator to evaluate an expression, you know whether or not you are looking for a real or complex result. When a user types `(-8)^(2/3)` and expects to get `-4`, the user is likely only interested in real results. There will always be `nthRoot` to fall back on if a user wants to pick and choose when to allow a complex result.

`nthRoot`, if I remember correctly, always returns a real number if both arguments are real, so the optional third argument isn't necessary.

**Comment by josdejong** [comment]
> In this way it is similar to predictable like you said, but not exactly, since it only suppresses the output of complex numbers. I think this would be acceptable.

You're right, this makes sense. Also, if you *do* want complex valued results anyways you could always convert the input to complex first, like `pow(complex(x))`, I hadn't thought of that before. Ok then let's go for using the global option `predictable` for this!

As for `nthRoot`: I think that we can extend this function to both support negative numbers and a non-integer power and work similar to `pow` regarding real/complex valued output.

**Comment by mariussmit** [comment]
Glad to see you're discussing this. Came across the same issue today where math.eval("3^3^3") gave me 7625597484987 where I thought it would give me 19683. I expected the latter because that is what I get on a calculator as well as in Excel. I'm not a mathematician either so I don't know when which number is preferred. In my case though the latter answer seems to be the default one used in the general public space, if I can put it that way.
Plain JS: `Math.pow(Math.pow(3,3),3)`

**Comment by josdejong** [comment]
@mariussmit please open a new issue to discuss a new topic. You can find more about the behavior of `pow` in #498.

**Comment by mariussmit** [comment]
Awesome, thank you @josdejong.

**Comment by ericman314** [comment]
Please see PR #525. This causes `(-8)^(2/3)` to evaluate to `4` when `predictable:true` is set.

I haven't modified `nthRoot` yet, nor do I think it's entirely necessary. Matlab's documentation on `nthRoot(X,N)` states that if X is negative, then A must be an odd integer. Our implementation already works the same way.

**Comment by sonnyk22** [comment]
I think I am a little bit confused based on your reply.  Did you mean to say that by setting the predictable:true property on the .config() on the latest released version of MathJs, the expression of (-8)^(2/3) will evaluate to 4?

**Comment by ericman314** [comment]
Yes. In version 2.7.1 (not yet released) if you use `math.config({predictable: true})` then `(-8)^(2/3)` will return 4, the real root, rather than `-2 + 3.4641016151378i`, the principle root. Not every expression using `^` has a real root; for example, `(-8) ^ (1/2)` has only imaginary roots.

**Comment by sonnyk22** [comment]
Thank you for taking the time to implement this feature.  Would you say that the new version 2.7.1 will be released sometime in the next 2 weeks or so?

**Comment by josdejong** [comment]
It will become a v3.0.0 and I expect it indeed within a few weeks. If you need it very badly now you can always build it yourself from the develop branch and use that until the next version is released.

**Comment by josdejong** [comment]
v3 is now released, enjoy

**Comment by sonnyk22** [comment]
Awesome! Thank you!!!!!

**Comment by sonnyk22** [comment]
Quick question; Do I have some possible missing setting(s) in my .config() shown below to get the correct result for (-8)^(2/3)?
{epsilon: 1e-12, matrix: "Matrix", number: "BigNumber", precision: 64, predictable: true})

I have gotten the latest version of 3.0.

**Comment by ericman314** [comment]
Just installed version 3.0 as well. Perhaps the `^` operator isn't supported by BigNumber in predictable mode, on account of the new "real mode" behavior? I'm not sure as I've never used BigNumber much before. By removing the BigNumber option your expression does return -4 (with round-off error), but you probably want to keep the BigNumber option.

**Comment by sonnyk22** [comment]
That's what it was.  I set it from BigNumber to number.  Running the testing scripts right now, so far so good.  Thank you!
